### PR TITLE
Add support for nested interpolations in Override grammar

### PR DESCRIPTION
**Pull request by odelalleau**
## Motivation

Nested interpolations are coming in OmegaConf. The simple regexp currently found in Hydra's lexer won't work with them (`INTERPOLATION: '${' ~('}')+ '}';`).

Note: this PR is intended for next major version. I'm still putting it up for visibility -- I will try and keep it updated until it is time to merge it.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebookresearch/hydra/blob/master/CONTRIBUTING.md)?

Yes

## Test Plan

`pytest`

## Related Issues and PRs

Corresponding changes in OmegaConf: https://github.com/omry/omegaconf/pull/321
Older version of this PR: #879

### CODE REVIEW

File: hydra/core/override_parser/overrides_visitor.py [hunk 1]
```diff
@@ -173,6 +184,8 @@ def visitDictKeyValuePair(
     ) -> Tuple[str, ParsedElementType]:
         children = ctx.getChildren()
         item = next(children)
+        # We should never visit dictionaries that have interpolations as keys
+        # (currently these are only allowed within interpolations, which are not visited).
```
Comment (omry) [inline hydra/core/override_parser/overrides_visitor.py:188]: I don't understand this comment.
1. Since when do we have interpolation as keys in dictionaries?
2. is the code doing anything to address the requirements from this comment?

File: hydra/grammar/OverrideLexer.g4 [hunk 2]
```diff
@@ -69,4 +71,17 @@ QUOTED_VALUE:
...
     | '"' ('\\"'|.)*? '"' ;  // Double quotes, can contain escaped double quote : /"
 
-INTERPOLATION: '${' ~('}')+ '}';
+////////////////////////
+// INTERPOLATION_MODE //
+////////////////////////
+
+mode INTERPOLATION_MODE;
+
+NESTED_INTER_OPEN: INTER_OPEN -> type(INTER_OPEN), pushMode(INTERPOLATION_MODE);
```
Comment (omry) [inline hydra/grammar/OverrideLexer.g4:80]: not sure if this token type is needed.
looks like it's exactly like INTER_OPEN.
Is this because we need to handle all possible tokens in the mode?

File: hydra/grammar/OverrideParser.g4 [hunk 3]
```diff
@@ -47,12 +47,17 @@ function: ID POPEN (argName? element (COMMA argName? element )* )? PCLOSE;
...
 // Data structures.
 
-listValue: BRACKET_OPEN                          // [], [1,2,3], [a,b,[1,2]]
-    (element(COMMA element)*)?
-BRACKET_CLOSE;
-
+listValue: BRACKET_OPEN sequence? BRACKET_CLOSE;                          // [], [1,2,3], [a,b,[1,2]]
 dictValue: BRACE_OPEN (dictKeyValuePair (COMMA dictKeyValuePair)*)? BRACE_CLOSE;  // {}, {a:10,b:20}
-dictKeyValuePair: ID COLON element;
+dictKeyValuePair: (ID | interpolation) COLON element;
```
Comment (omry) [inline hydra/grammar/OverrideParser.g4:52]: I don't think interpolation in a key makes sense. can you show a use case?

File: hydra/grammar/OverrideParser.g4 [hunk 4]
```diff
@@ -47,12 +47,17 @@ function: ID POPEN (argName? element (COMMA argName? element )* )? PCLOSE;
...
+sequence: element (COMMA element)*;
+
+// Interpolations.
+
+interpolation: interpolationNode | interpolationResolver;
+interpolationNode: INTER_OPEN DOT* configKey (DOT configKey)* INTER_CLOSE;
```
Comment (omry) [inline hydra/grammar/OverrideParser.g4:58]: add tests for relative interpolations, I think it's not covered here.

File: news/976.feature [hunk 5]
```diff
@@ -0,0 +1 @@
+Override grammar now supports nested interpolations (ex: `python my_app.py 'key=${foo.${bar}}'`)
```
Comment (omry) [inline news/976.feature:1]: Bump the version of Hydra to 1.1.0-dev1.
This is the first major change.

File: tests/test_overrides_parser.py [hunk 6]
```diff
@@ -1953,3 +1954,309 @@ def test_sweep_iterators(
...
+# OmegaConf, and properly identify parsing error cases.
+# (should be replaced by imports from OmegaConf in the future)
+GrammarSyntaxError = HydraException
+# The OmegaConf exceptions below cannot be identified at parsing time.
+ConfigKeyError = GrammarTypeError = UnsupportedInterpolationType = None
+
+# DO NOT EDIT `TEST_CONFIG_DATA`!
+# It should be copied regularly from OmegaConf to ensure Hydra's grammar is able
+# to parse similar expressions as OmegaConf's.
+TEST_CONFIG_DATA: List[Tuple[str, Any, Any]] = [
```
Comment (omry) [inline tests/test_overrides_parser.py:1969]: this test file is already massive.
Can you put the OmegaConf tests in a different test file?

**Review by omry — CHANGES_REQUESTED** [review]

File: hydra/core/override_parser/overrides_visitor.py (hunk 1 above)
Comment (odelalleau) [reply hydra/core/override_parser/overrides_visitor.py:188]: I'll try to make this comment clearer if I can convince you to keep interpolations as keys ;) But to answer your questions:

1. Since https://github.com/omry/omegaconf/pull/321
2. That's actually the assert just below this comment (but I'm not entirely sure I understood your question)

For additional context:
- In OmegaConf, when writing the grammar I added support for dictionaries with interpolations as keys. Such dictionaries can only happen when inside an interpolation.
- When "merging" OmegaConf's grammar within Hydra, I wondered what to do if someone wanted to define a dictionary with interpolations as keys *outside* of an interpolation. To be honest I didn't spend much time thinking about it and decided just not to support it for now out of simplicity (thus keeping the assert below the comment).

File: hydra/grammar/OverrideLexer.g4 (hunk 2 above)
Comment (odelalleau) [reply hydra/grammar/OverrideLexer.g4:80]: All modes are entirely independent, if you want to re-use a rule from another mode it has to be declared again (under a different name)

File: news/976.feature (hunk 5 above)
Comment (odelalleau) [reply news/976.feature:1]: Done in 416abb1b1

File: tests/test_overrides_parser.py (hunk 6 above)
Comment (odelalleau) [reply tests/test_overrides_parser.py]: Good idea, done in 044a006f2

**Comment by omry** [comment]
cool , before I merge - can you squash the news file addition to the previous commit?

**Comment by omry** [comment]
in fact, make it two commits:
1. everything
2. the version bump

File: hydra/grammar/OverrideParser.g4 (hunk 4 above)
Comment (odelalleau) [reply hydra/grammar/OverrideParser.g4:58]: True, my big list of tests from OmegaConf doesn't have examples of nested interpolations. However I ran into an issue when adding some -- will need more time to look into it (this could have to wait until next week).

File: hydra/grammar/OverrideParser.g4 (hunk 3 above)
Comment (odelalleau) [reply hydra/grammar/OverrideParser.g4:52]: Well I don't see a reason why it would make more sense to allow interpolations in values compared to keys, for a start.

First thing I can think of (probably not the most complete/simple/convincing), let's say there are different folders, each containing a known type of files, and each file must be processed by a function that depends on its type. And let's say we want to perform some kind of reduce operation over the processed files, whose output is saved in the config variable `reduced`. It may look like this:

```yaml
images:
    path: /path/to/images
    process_function: process_image
videos:
    path: /path/to/videos
    process_function: process_video
captions:
    path: /path/to/captions
    process_function: process_caption
reduced: ${process:{${images.path}: ${images.process_function}, ${videos.path}: ${videos.process_function}, ${captions.path}: ${captions.process_function}}}
```

The code for the `process` resolver would be something like:

```python
def process(**to_process):
    return some_reduce_operation(f(os.path.join(path, x)) for path, f in to_process.items() for x in os.listdir(path))
```

File: hydra/core/override_parser/overrides_visitor.py (hunk 1 above)
Comment (omry) [reply hydra/core/override_parser/overrides_visitor.py:188]: 1. Since I didn't really get to review it yet, I was not aware of this intention. can you show me a convincing use case for it?
2. The assert is the answer to my second question. the comment like an unrelated piece of information. I did not realize the assert was really about enforcing it.

**Review by omry — CHANGES_REQUESTED** [review]
1. I think we should drop interpolation in keys. this is way too crazy.
The key in a hashtable should be fixed. I am not sure how it can even work when an action outside of the key can change it at runtime (what happens to the hash of the key at that point in the hashtable which is map?)
You are welcome to change my mind.

2. Please squash the commits as requested (or alternatively undo the version bump, I can do it separately).

**Comment by odelalleau** [comment]
> 1. I think we should drop interpolation in keys. this is way too crazy.
>    The key in a hashtable should be fixed. I am not sure how it can even work when an action outside of the key can change it at runtime (what happens to the hash of the key at that point in the hashtable which is map?)
>    You are welcome to change my mind.

I won't try too hard to change your mind, since anyway a dict can always be replaced with a list. But your comment confuses me and makes me wonder if we're talking about the same thing, so I'll try to clarify:

This is only for dictionaries that are passed as inputs to resolvers. These dictionaries are re-created on-the-fly on every access to the variable calling such a resolver. So we're never modifying in-place the keys of a dictionary.

File: hydra/grammar/OverrideParser.g4 (hunk 3 above)
Comment (omry) [reply hydra/grammar/OverrideParser.g4:52]: Okay, I will comment on the other thread. (easier to find).

File: hydra/grammar/OverrideParser.g4 (hunk 4 above)
Comment (omry) [reply hydra/grammar/OverrideParser.g4:58]: it could be in a followup. relative interpolations are very new in OmeaConf.

**Review by omry — COMMENTED** [review]
Thanks for the use case.
One thing to remember is that the primary use case for dictionaries is to merge them into the config, where they will be accessible later.

```yaml
a:
  x: 10
  y: 20

# command line
a={a:{z:30}}

# result
a:
  x: 10
  y: 20
  z: 30
```
Now, assuming you have support for interpolation in keys, what does this do?
```yaml
a:
  x: 10
  y: 20
b: ???

# command line 1:
#
# this is evaluated and merged right after config composition, later modifications to
# a.x would not apply to it anymore, which is unexpected because interpolations 
# are normally evaluated lazily.

b: {foo_${a.x}: 30}

# result:
a:
  x: 10
  y: 20
b:
  foo_10: 30
```

Interpolations can also point to nodes (dicts or lists in the config), this is another things to consider - lists and dicts cannot be keys in OmegaConf configs.

```
a:
  x: 10
  y: 20
b: ???
# this is hopefully going to blow up with a nice error
b: {${a}: 10}
```

In short, before enabling interpolation into keys, I want tests covering the functionalities it enables.
In particular, OmegaConf.update() should be tested with dict values that has such keys to ensure it does the right thing, but there could be other areas that are affected.
If you want to try to support it, file an issue against OmegaConf and we can iterate on it independently from the rest of the changes.

**Comment by odelalleau** [comment]
> One thing to remember is that the primary use case for dictionaries is to merge them into the config, where they will be accessible later.

Ah I see! So actually what you describe is not supported, trying to use an interpolation as key outside of an interpolation will fail on the assert I pointed out earlier (https://github.com/odelalleau/hydra/blob/044a006f24268a2991a0c37f1bcaca4a40e6621c/hydra/core/override_parser/overrides_visitor.py#L189)

As I mentioned above, _"this is only for dictionaries that are passed as inputs to resolvers"_.

At its source, the problem lies on the assumption that it's ok to use the same grammar to parse command-line overrides that set the value of a config option and to parse resolver arguments. This is helpful to keep Hydra's and OmegaConf's grammar close to each other, but it's not surprising to hit a few edge cases where things don't match perfectly.

Right now I see 3 options:

1. Keep things as they are in the current version of this PR (as I said, the situation you mentioned is currently explicitly not supported)
2. Get rid of interpolations in keys of dictionaries passed as resolver arguments (side note: this has the added benefit of simplifying OmegaConf's code -- this is related to changes in my OmegaConf's PR that you haven't gone through yet)
3. Better separate Hydra and OmegaConf grammars so that they can both evolve separately

As of now my preference is for (1) since it means less work for me, but I don't feel strongly about it so let me know what you prefer.

**Comment by omry** [comment]
> > One thing to remember is that the primary use case for dictionaries is to merge them into the config, where they will be accessible later.
> 
> Ah I see! So actually what you describe is not supported, trying to use an interpolation as dictionary key outside of an interpolation will fail on the assert I pointed out earlier (https://github.com/odelalleau/hydra/blob/044a006f24268a2991a0c37f1bcaca4a40e6621c/hydra/core/override_parser/overrides_visitor.py#L189)

Generally assertions are not a good way to tell users that they are doing something wrong, but that's besides the point.
Good that you (accidentally?) prevented that :).

> As I mentioned above, _"this is only for dictionaries that are passed as inputs to resolvers"_.
> 
> At its source, the problem lies on the assumption that it's ok to use the same grammar to parse command-line overrides that set the value of a config option and to parse resolver arguments. This is helpful to keep Hydra's and OmegaConf's grammar close to each other, but it's not surprising to hit a few edge cases where things don't match perfectly.
> Right now I see 3 options:
> 
> 1. Keep things as they are in the current version of this PR (as I said, the situation you mentioned is currently explicitly not supported)
> 2. Get rid of interpolations in keys of dictionaries passed as resolver arguments (side note: this has the added benefit of simplifying OmegaConf's code -- this is related to changes in my OmegaConf's PR that you haven't gone through yet)
> 3. Better separate Hydra and OmegaConf grammars so that they can both evolve separately
> 
> As of now my preference is for (1) since it means less work for me, but I don't feel strongly about it so let me know what you prefer.

Do you have a concrete use case for this feature or is this theoretical?
my preference is to introduce it separately (if if there are compelling use cases) and not as part of the initial OmegaConf grammar because I am not convinced that there are not unforeseen implications to adding it yet.

Let's go with 2 for now.

**Comment by odelalleau** [comment]
> Let's go with 2 for now.

Alright, I'll make this PR as draft until I can change it and also look into that other issue related to relative interpolations. ETA sometime next week after ICLR deadline.

**Comment by omry** [comment]
> > Let's go with 2 for now.
> 
> Alright, I'll make this PR as draft until I can change it and also look into that other issue related to relative interpolations. ETA sometime next week after ICLR deadline.

Sounds good. hopefully by that time I will have time to start going over your big PR at the OmegaConf repo.

**Comment by odelalleau** [comment]
Pushed new version, should be ok now:
* Squashed commits
* Added test for relative interpolations (this is in `test_omegaconf_grammar.py`)
* Interpolations aren't allowed anymore as keys in dictionaries passed as resolver arguments

**Comment by omry** [comment]
Now that Hydra is on top of OmegaConf 2.1.0dev20 that contains the nested interpolation support in OmegaConf, this is unblocked.
Looks like there are some conflicts.

When you think the time is right, we can work toward integrating this.

**Comment by odelalleau** [comment]
> Now that Hydra is on top of OmegaConf 2.1.0dev20 that contains the nested interpolation support in OmegaConf, this is unblocked.
> Looks like there are some conflicts.
> 
> When you think the time is right, we can work toward integrating this.

Yes, I was planning to get back to it once my current work on OmegaConf would be finalized. There are definitely some changes to make to this PR, based on how things have landed in OmegaConf.

**Comment by odelalleau** [comment]
Superseded by #1594 => closing
